src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -295,6 +295,7 @@ public void DiscardBufferedData()
             _byteLen = 0;
             _charLen = 0;
             _charPos = 0;
+            _checkPreamble = _encoding.Preamble.Length > 0;
```

Comment:
Hello Stephen!

If the stream is not seekable, why would the developer want to call `DiscardBufferedData`? That method should only be used when the underlying stream's position has been changed?

Right now, `StreamReader` detects BOM at any point in the stream, and skips it **as long as** it's the first time that `StreamReader` instance has read data. If the flag for detecting the encoding via BOM was enabled, that also means setting the new encoding to the read BOM. 

I took that already established notion and made `DiscardBufferedData` to act as if the consumer had just started using a new instance of the `StreamReader` class, with the only difference that it won't auto-detect the encoding, which seems like an inconsistency to me.

>I don't think this is correct or sufficient. What if, for example, the developer is using this method because they want to >seek back just a few bytes or forward to a known location rather than to the beginning of the stream?

Yes, you're right. This introduces a breaking change, in that previously if you had a stream with multiple BOMs, using `DiscardBufferedData` would never consume a BOM (be it at the start, or somewhere else), if you've changed the position of the underlying stream to the beginning of one such BOM.

Now, is it a good idea to consume a BOM (and potentially detect the new encoding) after resetting the buffer? Maybe? Should we expect that a reset to the buffer puts the `StreamReader` class in a state as if it had just been instantiated, so that the next read it does, it skips any BOM found, and if enabled, auto detects the encoding?

Reply:
Hi :)

> If the stream is not seekable, why would the developer want to call DiscardBufferedData? That method should only be used when the underlying stream's position has been changed? Is it in cases where something else is reading from it?

Developers do amazingly unexpected things all the time.  We can't predict all the ways libraries are used.

> with the only difference that it won't auto-detect the encoding, which seems like an inconsistency to me

I don't think it's an inconsistency.  I think it's an acknowledgement on the part of the developer that wrote this that it's not possible to correctly implement BOM re-detection in this method as it's defined.  The method is about throwing away buffered data, nothing more.

> Should we expect that a reset to the buffer puts the StreamReader class in a state as if it had just been instantiated?

I think we should accept the existing behavior for the existing method.

Reply:
I agree. If anything, this seems like a feature request, allowing a single `StreamReader` to handle multiple encodings, and for that to happen, we'd have to introduce surface API changes. Do you think a method that resets the encoding/BOM is a meaningful addition to the `StreamReader`, or vast majority of people should just create a new instance of it?

Reply:
> Do you personally think a method that resets the encoding/BOM is a meaningful addition to the StreamReader, or vast majority of people are better of creating a new instance of it?

Personally, I've never had a need, but that doesn't mean one doesn't exist.  Someone would need to do the research on its worth.